File: src/main/kotlin/com/github/magneticflux/aoc/day3/Problem2.kt
package com.github.magneticflux.aoc.day3

import com.github.magneticflux.aoc.input
import com.github.magneticflux.aoc.spiralSequenceOf
import com.github.magneticflux.geom.IntPoint2D
import java.util.*

/**
 * Created by Mitchell Skaggs on 12/1/2017.
 */
object Problem2 {
    @JvmStatic
    fun main(args: Array<String>) {
        val input = Scanner(input())
        val location = input.next().toInt()

        val map = mutableMapOf(IntPoint2D(0, 0).to(1))

        val firstLarger = spiralSequenceOf()
                .drop(1) // Exclude first tile
                .find {
                    val value = it.surroundingPoints.map { map[it] ?: 0 }.sum()

                    map[it] = value

                    return@find value > location
                }
        println("First larger than $location at $firstLarger = ${map[firstLarger]}")
    }
}

File: src/main/scala/com/github/magneticflux/aoc/year2018/day2/Problem2.scala
package com.github.magneticflux.aoc.year2018.day2

import com.github.magneticflux.aoc.InputHelpers._

/**
  * Created by Mitchell Skaggs on 12/2/2018.
  */
object Problem2 {
  def main(args: Array[String]): Unit = {
    val lines = this.input()
      .getLines()
      .toSeq

    val pair = lines.combinations(2)
      .map { case Seq(s1, s2) => (s1, s2) }
      .find(p => {
        val (s1, s2) = p

        val numDiffs = s1.zip(s2)
          .map(p => {
            if (p._1 != p._2) 1 else 0
          })
          .sum

        numDiffs == 1
      })
      .get

    println(
      pair._1.zip(pair._2).filter(p => p._1 == p._2).map(_._1).mkString
    )
  }
}

